core/store/src/migrations.rs

```diff
@@ -358,3 +360,39 @@ pub fn migrate_39_to_40(store: &Store) -> anyhow::Result<()> {
     update.commit()?;
     Ok(())
 }
+
+/// Migrates the database from version 40 to 41.
+///
+/// This rewrites the contents of the StateDlInfos column
+pub fn migrate_40_to_41(store: &Store) -> anyhow::Result<()> {
+    #[derive(BorshSerialize, BorshDeserialize)]
+    struct LegacyStateSyncInfo {
+        sync_hash: CryptoHash,
+        shards: Vec<ShardInfo>,
+    }
+
+    let mut update = store.store_update();
+
+    for row in store.iter_prefix_ser::<LegacyStateSyncInfo>(DBCol::StateDlInfos, &[]) {
+        let (key, LegacyStateSyncInfo { sync_hash, shards }) =
+            row.context("failed deserializing legacy StateSyncInfo in StateDlInfos")?;
+
+        let epoch_first_block = CryptoHash::try_from_slice(&key)
+            .context("failed deserializing CryptoHash key in StateDlInfos")?;
+
+        if epoch_first_block != sync_hash {
+            tracing::warn!(key = %epoch_first_block, %sync_hash, "sync_hash field of legacy StateSyncInfo not equal to the key. Something is wrong with this node's catchup info");
+        }
+        let new_info = StateSyncInfo {
+            state_sync_version: 0,
+            epoch_first_block,
+            sync_hash: Some(sync_hash),
+            shards,
+        };
+        update
+            .set_ser(DBCol::StateDlInfos, &key, &new_info)
+            .context("failed writing to StateDlInfos")?;
```

Comment:
Sanity check - is there any potential for this update to OOM?

Reply:
I guess I just don't know how does cleanup work for those - if you happen to know please share :)

Reply:
I think in practice there's really only going to be one or zero keys in the `StateDlInfos` column, since the key is the hash of the first block of an epoch, and it's cleaned up after catchup is done, so really we only see one iteration of this loop in practice